core/sail/base/src/main/java/org/eclipse/rdf4j/sail/base/SailSourceConnection.java

```diff
@@ -614,12 +614,23 @@ public boolean removeInferredStatement(Resource subj, IRI pred, Value obj, Resou
 
 	private boolean remove(Resource subj, IRI pred, Value obj, SailDataset dataset, SailSink sink, Resource... contexts)
 			throws SailException {
+
+		if (!hasConnectionListeners() && sink instanceof SailSinkVersion2
+				&& getIsolationLevel().isCompatibleWith(IsolationLevels.NONE)) {
+			return ((SailSinkVersion2) sink).deprecateByQuery(subj, pred, obj, contexts);
+		}
```

Comment:
Once you fold SailSinkVersion2 into the "normal" SailSink, this code can just go. Same with the bit on lines 629 and further.

Reply:
I just realized that part of the problem is that this private method seems to assume deprecate only accepts a single triple at a time, and does a dataset retrieval over which it iterates to remove individual triples. I can see how that could conceivably be inefficient. From what I can tell though, the reason we do that is not that `deprecate` doesn't accept wildcards (it does), but that we need a list of the statements that are removed to send to the listener.

Reply:
Yes. It’s that we need a list of statements for the listener and that we need to return true if a statement was removed. 

The code:

```
if (!hasConnectionListeners() && getIsolationLevel().isCompatibleWith(IsolationLevels.NONE)
 				&& !(sink instanceof Changeset)) {
```

 checks if there is a listener, that we are in isolation level null (which might be a bit too restrictive) and that the sink is not a changeset, because  changeset doesn’t support deprecatebyquery.

Reply:
Btw there is a bug in GitHub. It’s showing some old code in the code window at the start of this chain of comments.